File: LeetCodePy/FindIfPathExistsInGraph_1971.py
from typing import List

class FindIfPathExistsInGraph_1971:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        if source == destination:
            return True

        # Create adjacency list
        graph = [[] for _ in range(n)]
        for edge in edges:
            graph[edge[0]].append(edge[1])
            graph[edge[1]].append(edge[0])

        # Use a list to mark visited nodes
        visited = [False] * n
        visited[source] = True

        return self.dfs(source, destination, graph, visited)

    def dfs(self, node: int, destination: int, graph: List[List[int]], visited: List[bool]) -> bool:
        if node == destination:
            return True

        for neighbor in graph[node]:
            if not visited[neighbor]:
                visited[neighbor] = True
                if self.dfs(neighbor, destination, graph, visited):
                    return True

        return False

File: LeetCodeJava/FindIfPathExistsInGraph_1971.java
import java.util.ArrayList;
import java.util.List;

public class FindIfPathExistsInGraph_1971 {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        if (source == destination) return true;

        // Create an adjacency list using a list of lists
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            graph.get(edge[0]).add(edge[1]);
            graph.get(edge[1]).add(edge[0]);
        }

        // Use an array to mark visited nodes
        boolean[] visited = new boolean[n];
        visited[source] = true;

        return dfs(source, destination, graph, visited);
    }

    private boolean dfs(int node, int destination, List<List<Integer>> graph, boolean[] visited) {
        if (node == destination) return true;

        for (int neighbor : graph.get(node)) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                if (dfs(neighbor, destination, graph, visited)) return true;
            }
        }
        return false;
    }
}
